lib/engine/game/g_1862/game.rb

```diff
@@ -888,6 +935,442 @@ def check_intersection(routes)
+              max_stops = p
+            end
+          end
+          max_stops
+        end
+
+        def optimize_stops(route, _num_pay, _total_stops)
+          @global_stops ||= global_optimize(route.routes)
+
+          @global_stops[route.routes.index(route)]
+        end
+
+        def compute_local_stops(route, visits)
+          if mn_train?(route.train)
+            towns = visits.select(&:town?)
+            optimize_stops(route) + towns
+          else
+            visits
+          end
+        end
+
+        def compute_express_stops(route, visits)
+          if mn_train?(route.train)
+            optimize_stops(route)
+          else
+            visits.select { |n| n.city? || n.offboard? }
+          end
+        end
+
+        def compute_stops(route)
+          @cached_freight_sets = nil
+          @global_stops = nil
+          visits = route.visited_stops
+          case train_type(route.train)
+          when :local
+            compute_local_stops(route, visits)
+          when :express
+            compute_express_stops(route, visits)
+          else
+            [visits.first, visits.last]
+          end
+        end
+
+        # FIXME: take options into account
+        def nonpermanent_freight?(train)
+          train.distance < 6
+        end
+
+        # given a route, and a list of available stops,
+        # recursively find a set of other routes that connect to it end-to-end
+        # when faced with a branch, choose:
+        # 1. the branch containing non-permanent trains, then
+        # 2. the longest branch
+        #
+        def get_set_node(node, route, stops)
+          return [] unless stops
+
+          stops.delete(route)
+          return [] if stops.empty? # no stops left, we're done
+
+          routes = stops.keys.select { |r| stops[r].include?(node) }
+          if routes.size == 1
+            # no branching, continue recursing
+            create_oneway_set(node, routes[0], [routes[0]], stops)
+          elsif routes.size > 1
+            # branching, recurse on all branches and pick
+            branches = []
+            routes.each do |r|
+              new_stops = stops.dup
+              branches << create_oneway_set(node, r, [], new_stops)
+            end
+            branch = if (nonperm_set = branches.find { |s| s.any? { |r| nonpermanent_freight?(r.train) } })
+                       nonperm_set
+                     else
+                       branches.max_by(&:size)
+                     end
+            branch.each { |r| stops.delete(r) }
+          else
+            # no other routes match, we're done
+            []
+          end
+        end
+
+        def create_oneway_set(visited, route, set, stops)
+          set << route
+          end_a = stops[route].first
+          end_b = stops[route].last
+          set.concat(get_set_node(end_a, route, stops)) if end_a != visited
+          set.concat(get_set_node(end_b, route, stops)) if end_b != visited
+          set.compact.uniq
```

Comment:
why will there be nils here or dups? if this is a set there would be no dups, but this must not be a set

Reply:
I left these in due to paranoia. You are right, these are not true Ruby sets. I will experiment to make sure they aren't needed. In the "normal" case they aren't, but I was worried about being fed degenerate routes.

Reply:
Yes, both are needed. There can be dups and nils.